File: graph/거리두기확인하기.cpp
#include <string>
#include <queue>
#include <vector>

using namespace std;

int dx[4] = {-1, 0, 1, 0},
    dy[4] = {0, -1, 0, 1};

struct point{
    int x, y;
};

bool checkPlaces(vector<string> &place) {
    for(int j = 0; j < place.size(); j++){
        for(int k = 0; k < place[j].size(); k++){
            if(place[j][k] == 'P'){
                bool visited[5][5] = {false, };
                queue<pair<point, int>> q;
                q.push({{j, k}, 0}); visited[j][k] = true;
                while(!q.empty()){
                    pair<point, int> x = q.front(); q.pop();
                    if(x.second >= 2) continue;
                    for(int z = 0; z < 4; z++){
                        int xx = x.first.x + dx[z];
                        int yy = x.first.y + dy[z];
                        if(xx > -1 && xx < 5 && yy > -1 && yy < 5 && !visited[xx][yy]){
                            if(place[xx][yy] == 'O'){
                                q.push({{xx, yy}, x.second + 1}); visited[xx][yy] = true;
                            } else if (place[xx][yy] == 'P'){
                                return false;
                            }
                        }
                    }
                }
            }
        }
    }
    return true;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for(int i = 0; i < places.size(); i++){
        if(checkPlaces(places[i])) answer.push_back(1);
        else answer.push_back(0);
    }
    return answer;
}

int main(){
    solution({{"POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"}, {"POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"}, {"PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"}, {"OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"}, {"PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"}});
    return 0;
}

//[["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]

File: graph/거리두기확인하기.swift
//
//  main.swift
//  Swift-PS
//
//  Created by 김민창 on 2021/07/24.
//

import Foundation


func solution(_ places:[[String]]) -> [Int] {
    let dx = [-1, 0, 1, 0], dy = [0, -1, 0, 1]
    var result = [Int]()
    
    places.forEach{
        var visited = Array(repeating: Array(repeating: false, count: 5), count: 5)
        var keepDistance = true
        for i in 0..<$0.count {
            for j in 0..<$0[i].count {
                if $0[i][j] == "P" {
                    var q = [(Int, Int, Int)]()
                    
                    q.append((i, j ,0));
                    visited[i][j] = true
                    
                    while !q.isEmpty {
                        let before = q.removeFirst()
                        if before.2 >= 2 {
                            continue
                        }
                        for i in 0..<4 {
                            let xx = before.0 + dx[i]
                            let yy = before.1 + dy[i]
                            if 0...4 ~= xx && 0...4 ~= yy {
                                if $0[xx][yy] == "X" || visited[xx][yy] {
                                    continue
                                }
                                if $0[xx][yy] == "P" {
                                    keepDistance = false
                                    q.removeAll()
                                    break
                                }
                                
                                q.append((xx, yy, before.2 + 1))
                            }
                        }
                    }
                    if !keepDistance {
                        break
                    }
                }
                if !keepDistance {
                    break
                }
            }
            if !keepDistance {
                break
            }
        }
        if keepDistance {
            result.append(1)
        } else {
            result.append(0)
        }
    }
    
    return result
}

extension String {
    subscript(_ index: Int) -> Character {
        return self[self.index(self.startIndex, offsetBy: index)]
    }
}


print(solution([["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]]))
